### ISSUE TITLE
No Progress Statistics on Google Colab

### ISSUE THREAD

**Issue by SirNeural** [fastai/fastai#899]
**Describe the bug**
When running a pretrained ConvLearner on Google Colab(haven't tested with other learners), and doing lr_find, fit, and fit_one_cycle, no progress statistics are printed until the epoch is over. 

**To Reproduce**

```
from fastai import *
from fastai.vision import *
data = image_data_from_folder('data/dogscats', ds_tfms=get_transforms(), tfms=imagenet_norm, size=224, num_workers=1, bs=32)
learn = ConvLearner(data, tvm.resnet34, metrics=accuracy, pretrained=True)
learn.lr_find()
learn.recorder.plot()
```

**Expected behavior**
Progress statistics, such as iterations per second, % of epoch completed, and rolling updates of loss, accuracy, etc with a progress bar should be printed.

**Screenshots**

https://i.imgur.com/fRDR4B1.png
https://i.imgur.com/ydSPXlC.png

**Additional context**
Running on Google Colab in a Python 3 Notebook with GPU backend; I found a forum post about progress bars missing, [Progress Bars in iPython Notebooks](http://forums.fast.ai/t/progress-bars-in-ipython-notebooks/22826/19), and tried the workaround described, 
```
from fastprogress import force_console_behavior
master_bar, progress_bar = force_console_behavior()
basic_train.master_bar, basic_train.progress_bar = master_bar, progress_bar
```
After running, the problem still persists. I've also tried resetting the runtime multiple times.

**Comment by sgugger** [comment]
If the stats aren't showing, it's because you don't have ipywidgets properly installed. You should check you enabled it properly. For the workaround, I think it doesn't work because you should type
```
from fastprogress import force_console_behavior
master_bar, progress_bar = force_console_behavior()
fastai.basic_train.master_bar, fastai.basic_train.progress_bar = master_bar, progress_bar
```

**Comment by SirNeural** [comment]
Hi @sgugger ,
I'm importing fastai in my code as 
```
from fastai import *
```
and if I try running fastai.basic_train.master_bar, etc, i get the following error: 
```
NameError                                 Traceback (most recent call last)
<ipython-input-15-1b7fc7d5e975> in <module>()
      1 from fastprogress import force_console_behavior
      2 master_bar, progress_bar = force_console_behavior()
----> 3 fastai.basic_train.master_bar, fastai.basic_train.progress_bar = master_bar, progress_bar

NameError: name 'fastai' is not defined
```

I've tried modifying my code to just
```
import fastai
```
but it still doesn't show any statistics https://i.imgur.com/thjyvdt.png.
From what I've read online, I don't think Google Colab supports ipywidgets, but if I'm wrong, how would I enable it?

**Comment by sgugger** [comment]
Hum, if Google Colab doesn't support ipywidget, it should default automatically to the console behavior. What version of fastprogress do you have?
And yes, you should add `import fastai` before executing the block of three lines, if you didn't do it before. The fact it doesn't print anything is weird. Can you try running the basic vision example (that should be fast) to see if it's just waiting for the end of an epoch to print anything?

**Comment by SirNeural** [comment]
I have tried adding `import fastai` alone as well as adding an `import fastai` call after my `from fastai import *` call, both tries still result in no statistics being shown.
I am unable to access the __version__ property for fastprogress, but I'm installing fastai from the recommended steps in the readme, and the setup.py lists the fastprogress version requirements as >=0.1.10.
If I let it run, at the end of the epoch, it does print the statistics epoch by epoch, example (for dogs vs cats): https://i.imgur.com/dYRYLt6.png.

Running the vision example results in an undefined function error:
```
name 'untar_data' is not defined
```

**Comment by sgugger** [comment]
That last error comes from your version yes, it'll be fixed in a more recent one.
Can you try the basic example that is shown on the [fastpogress repo](https://github.com/fastai/fastprogress#usage) and tell me what you see?

**Comment by SirNeural** [comment]
Running the fast progress example repo with a pip install fastprogress at the top of the notebook results in just text being printed saying “Finished loop x”
Here the notebook for reference https://drive.google.com/open?id=1SuMS0UDi0g50fRFsHwYf4_IakNxSiI2S

**Comment by sgugger** [comment]
Thanks for your reports. The basic issue is
1. ipywidgets isn't supported on colab so we can't have the pretty progress bar
2. for the console behavior, print('foo', end='\r') doesn't behave as in any other terminal and erases the output instantly after printing it, which is why you don't see the console progress bar.

I'll try to see how to get around that (I see that tqdm works properly for instance) but there is a lot to do for the release of the course, so it might not be soon...

**Comment by SirNeural** [comment]
Alright thanks, just let me know if a workaround for this is found, since it would be nice to see live updates while waiting for the model to train.

**Comment by swang373** [comment]
Hi @sgugger , from googlecolab/colabtools#166, it appears that the behaviour of the carriage return in Colab also clears any output back to the newline, so... one solution could be a new subclass of ProgressBar that uses the output utilities provided by colabtools.

```python
# When on Google Colab, import the output utilities
try:
  from google.colab import output as colab_output
except ImportError:
    pass

# This is just a minimally modified version of ConsoleProgressBar
class ColabProgressBar(ProgressBar):
    length:int=50
    fill:str='█'

    def __init__(self, gen, total=None, display=True, leave=True, parent=None, auto_update=True):
        self.max_len,self.prefix = 0,''
        super().__init__(gen, total, display, leave, parent, auto_update)

    def on_iter_end(self):
        if not self.leave and printing():
            # On Google Colab, printed outputs can be marked with tags
            # using a (non-thread safe) context manager. Output can then
            # be selectively cleared by their tags. The clear() function can
            # also wait to clear output, deferring action until there is new output.
            # The following three lines first clears output tagged as 'fastprogress',
            # if any, and then prints output under that same tag. Note that the call
            # to print no longer terminates the line with a carriage return (end='\r')
            # because Colab would otherwise clear the contents we want to print.
            colab_output.clear(output_tags='fastprogress', wait=True)
            with colab_output.use_tags('fastprogress'):
                print(f'\r{self.prefix}' + ' ' * (self.max_len - len(f'\r{self.prefix}')))

    def on_update(self, val, text):
        if self.display:
            filled_len = int(self.length * val // self.total)
            bar = self.fill * filled_len + '-' * (self.length - filled_len)
            to_write = f'\r{self.prefix} |{bar}| {text}'
            if len(to_write) > self.max_len: self.max_len=len(to_write)
            if printing():
                # The same construct as above appears here as well,
                # and that's all of the modifications needed to get
                # a passable progress bar to show up on Colab
                colab_output.clear(output_tags='fastprogress', wait=True)
                with colab_output.use_tags('fastprogress'):
                    print(to_write)
```

If such a class were added to the fastprogress module, the `isnotebook()` function might need to be modified because when the module checks for the shell environment on Colab the value of `get_ipython().__class__.__name__` is just `'Shell'`, so it returns False rather than True. In any case, `IN_NOTEBOOK` would need to be True in order for `printing()` to return True when using Colab, otherwise the progress bar isn't printed.

This solution isn't perfect though, because some output lingers after the last iteration. For example, when I run the simple example featured in the README of fastprogress, I see the following output on Colab:
```
Finished loop 7.
Finished loop 8.
Epoch 10/10 :                                                                                                   
Finished loop 9.
```

The "Epoch 10/10" goes away if another call to `clear()` any outputs tagged as `'fastprogress'` is added just before the master bar writes out which loop it finished.

Hope this was helpful, cheers!

**Comment by sgugger** [comment]
Thanks for your suggestion. I'll indeed write a colab version of the bar when I have a bit of time. I did notice the ouput option, but there is also some way to make it a bit prettier: ipywidgets don't work fully but I found some HTML progress bars that rendered well on colab.

One thing that would be useful is to already have a test of whether or not the notebook is executed in colab (like IS_NOTEBOOK but IS_COLAB). For now I'm thinking to try 
```
from google.colab import output
```
to determine if I should set it True or False, but if you think of a better way, I'm all ears.

**Comment by swang373** [comment]
Nice, thanks for your hard work! I look forward to the pretty progress bars!

I think that check is fine, maybe one level up in case the submodules are shuffled around? 
```python
from google import colab
```
But that's more or less what you suggested. ;]

**Comment by gshashank84** [comment]
There is an AttributeError coming while running examples, which is given below
`AttributeError: module 'PIL' has no attribute 'Image'`

The installed package is 
`!pip show Pillow`
~~~
Name: Pillow
Version: 5.3.0
Summary: Python Imaging Library (Fork)
Home-page: http://python-pillow.org
Author: Alex Clark (Fork Author)
Author-email: aclark@aclark.net
License: Standard PIL License
Location: /usr/local/lib/python3.6/dist-packages
Requires: 
Required-by: fastai
~~~
It seems like the Pillow package of 5.3.0 does not support the Fastai V1

**Comment by sgugger** [comment]
This should be fixed by fastprogress 1.0.12.

**Comment by swang373** [comment]
Just letting you know that I've tried out the update in a Colab notebook and it looks great. Thanks for the new feature! 🍻
